src/Workspaces/CoreTestUtilities/MEF/TestComposition.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition.Hosting;
+using System.Reflection;
+using System.Text;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.Test.Utilities
+{
+    public sealed class TestComposition
+    {
+        public readonly ImmutableHashSet<Assembly> Assemblies;
+        public readonly ImmutableHashSet<Type> Types;
+
+        private readonly bool _vsMef;
+        private readonly Lazy<ContainerConfiguration> _lazyContainerConfiguration;
+        private readonly Lazy<IExportProviderFactory> _exportProviderFactory;
+
+        internal TestComposition(ImmutableHashSet<Assembly> assemblies, ImmutableHashSet<Type> types, bool vsMef)
+        {
+            Assemblies = assemblies;
+            Types = types;
+
+            _vsMef = vsMef;
+            _lazyContainerConfiguration = new Lazy<ContainerConfiguration>(() => new ContainerConfiguration().WithAssemblies(Assemblies).WithParts(Types));
+            _exportProviderFactory = new Lazy<IExportProviderFactory>(() => ExportProviderCache.GetOrCreateExportProviderFactory(GetCatalog()));
+        }
+
+        /// <summary>
+        /// Returns a new instance of <see cref="HostServices"/> for the composition. This will either be a MEF composition or VS MEF composition host, 
+        /// depending on what layer the composition is for. Editor Features and VS layers use VS MEF composition while anything else uses System.Composition.
+        /// </summary>
+        public HostServices GetHostServices()
+            => _vsMef ? GetVisualStudioHostServices() : (HostServices)new MefHostServices(_lazyContainerConfiguration.Value.CreateContainer());
+
+        internal VisualStudioMefHostServices GetVisualStudioHostServices()
+            => VisualStudioMefHostServices.Create(ExportProviderFactory.CreateExportProvider());
+
+        /// <summary>
+        /// VS MEF <see cref="ExportProvider"/>.
+        /// </summary>
+        public IExportProviderFactory ExportProviderFactory => _exportProviderFactory.Value;
+
+        private ComposableCatalog GetCatalog()
+            => ExportProviderCache.GetOrCreateAssemblyCatalog(Assemblies, ExportProviderCache.CreateResolver()).WithParts(Types);
+
+        public TestComposition WithAdditionalParts(params Type[]? types)
```

Comment:
What's the point of calling `WithAdditionalParts` with null for the `types` array?

Reply:
Some tests have helpers that allow to pass a list of additional parts in and the helper then creates the composition. The helper might have the array as an optional parameter.